core/common/src/LocalDateRange.kt

```diff
@@ -0,0 +1,118 @@
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as LocalDateProgression
+
+        return intProgression == other.intProgression
+    }
+
+    override fun hashCode(): Int {
+        return intProgression.hashCode()
+    }
+
+
+    public companion object {
+        public fun fromClosedRange(
+            rangeStart: LocalDate,
+            rangeEnd: LocalDate,
+            stepValue: Int,
+            stepUnit: DateTimeUnit.DayBased
+        ): LocalDateProgression = LocalDateProgression(rangeStart, rangeEnd, stepValue * stepUnit.days)
+    }
+}
+
+public class LocalDateRange(start: LocalDate, endInclusive: LocalDate) : LocalDateProgression(start, endInclusive, 1), ClosedRange<LocalDate>, OpenEndRange<LocalDate> {
+    override val start: LocalDate get() = first
+    override val endInclusive: LocalDate get() = last
+    override val endExclusive: LocalDate get() = endInclusive.plus(1, DateTimeUnit.DAY)
+
+    @Suppress("ConvertTwoComparisonsToRangeCheck")
+    override fun contains(value: LocalDate): Boolean = first <= value && value <= last
+
+    override fun isEmpty(): Boolean = first > last
+
+    override fun toString(): String = "$first..$last"
+
+    public companion object {
+        private val DATE_ONE = LocalDate(1, 1, 1)
+        private val DATE_TWO = LocalDate(1, 1, 2)
+        public val EMPTY: LocalDateRange = LocalDateRange(DATE_TWO, DATE_ONE)
+    }
+}
+
+public fun LocalDateProgression.first(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.first
+}
+public fun LocalDateProgression.last(): LocalDate {
+    if (isEmpty())
+        throw NoSuchElementException("Progression $this is empty.")
+    return this.last
+}
+public fun LocalDateProgression.firstOrNull(): LocalDate? = if (isEmpty()) null else this.first
+public fun LocalDateProgression.lastOrNull(): LocalDate? = if (isEmpty()) null else this.last
+
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(intProgression.reversed())
+
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(intProgression.step(value * unit.days))
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toInt(), unit)
+
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+public infix fun LocalDate.downUntil(that: LocalDate) : LocalDateProgression = downTo(that.plus(1, DateTimeUnit.DAY))
+
+public operator fun LocalDate.rangeTo(that: LocalDate): LocalDateRange = LocalDateRange(this, that)
+public operator fun LocalDate.rangeUntil(that: LocalDate) : LocalDateRange = rangeTo(that.minus(1, DateTimeUnit.DAY))
```

Comment:
This can overflow and throw, and this can be avoided.

Reply:
Looking to understand how best to avoid the overflow. If I'm understanding correctly, this can only overflow if `that` is `LocalDate.MIN` (every other date has a valid `LocalDate` one day before it). By definition this would make the range empty, as every possible `this` that yields a non-empty range would itself be less than `LocalDate.MIN` and therefore not valid. This would seem to imply you can just return `LocalDateRange.EMPTY` any time `that` is `LocalDate.MIN`, however `EMPTY` would have incorrect values for `start` and `endInclusive`, which are part of the public API. Even if you preserved `start` by returning `this.rangeTo(LocalDate.MIN)`, your `endInclusive` would still be one day off, and if `this` happened to be `LocalDate.MIN` you'd accidentally create a non-empty range. I'm not sure how to avoid the throw while still maintaining correctness in the resulting range.

Reply:
As a rule of thumb, when in doubt, just try to keep the behavior consistent with the stdlib:
```kotlin
fun main() {
    val range = 10 until (Int.MIN_VALUE)
    println(range) // prints "1..0"
}
```
So it's okay to return `EMPTY` if you notice you can't do anything meaningful and the range *is* empty. When it's possible to preserve the boundaries, then sure, it's worth doing:
```kotlin
fun main() {
    val range = 10 until (-1)
    println(range) // prints "10..-2"
}
```